### ISSUE TITLE
Option to autoscroll to the top of linked pages

### ISSUE THREAD

**Issue by anewuser** [mozilla/pdf.js#18229]
Configuration:
- Web browser and its version: Firefox stable
- Operating system and its version: Windows 10
- PDF.js version: -
- Is a browser extension: -

Steps to reproduce the problem:
1. Open a large medical PDF file with many figures and links to their labels
2. Click on one of the links

What is the expected behavior?

When a link points to a figure, users would like to see the figure right away.

What went wrong?

Some PDF authors place anchors to labels **below** the figures.

In other words, you click on "Figure 5" and just see its label. You have to scroll up to actually see the picture.

Researchers who read long PDF files become very frustrated with this situation.

![pdf](https://github.com/mozilla/pdf.js/assets/6733770/963441ed-2357-49ba-abb6-68b6c38fcef4)

Is there any solution to this issue? Maybe a way to ignore the exact location of the anchor and always show the top of the page it is on?

**Comment by Snuffleupagus** [comment]
While this would be easy enough to implement, given that it seems very much like a special-case it's just not clear to me that it's actually something that we want to support (since all features add complexity)!?
This is also, as far as I can remember, the first time something like this has been requested.

/cc @timvandermeij, @calixteman Thoughts?

**Comment by timvandermeij** [comment]
I'm also a bit on the fence about this. In principle this is the PDF generator's fault, and it's not entirely clear to me at least how we would be able to determine if we need to jump elsewhere or not (in other words: how do we know for sure that the label location is "incorrect"?). However, I think for me accepting this feature would depend on:

1. how much code complexity this would add;
2. if we can write tests for this;
3. how the functionality is triggered.

The latter might be a viewer option that interested users could set, but I think it should be off by default unless there is a way to be absolutely sure that the user intends to jump elsewhere, otherwise the annoyance described in the issue could also happen the other way around so to say for users that _did_ actually intend to jump to the exact location and are then redirected to the page top instead.

**Comment by calixteman** [comment]
Yep it'd nice to have a real pdf with a clear STR and then I can share with the UX team to have their inputs.

**Comment by Snuffleupagus** [comment]
The smallest patch I can imagine would be something along these lines: https://github.com/Snuffleupagus/pdf.js/compare/issue-18229 but it *obviously cannot* be enabled by default since that'd break so many more things than it'd potentially "fix".

**Comment by propylAlcohol** [comment]
In my experience, other PDF viewers such as those integrated in Adobe Acrobat and Chrome have similar behaviours to the current version of Firefox. This is for a good reason. The vast majority of bookmarks are intended to take you to the start of a paragraph or section where the content of interest is below the bookmark. The only exceptions to this that I am aware of are figures and graphs. For a typical novel, this would not be a problem, however, scientific textbooks and papers are very often littered with references to figures located multiple pages away. This requires you to constantly jump back and forth between the text and the figure to get an idea of what the text is trying to convey. I would argue for this reason that figure references are clicked/used much more often than regular bookmarks located in the PDF outline. 

Considering the gap in the market and the general sameness of pdf viewers I personally think that adding this feature would greatly increase Firefox's attractiveness for students like me in university or higher education.  

The most effective way to implement this would probably be to see if the clickable link has the word "Fig" in it and change the behaviour accordingly. 

Btw I have no idea how this would be implemented. You guys know best. Im not a programmer but a life sciences student. Thank you guys for your support on this thread :)

I also included an example of a paper which I am currently studying that could be used as a guide for the issue.
[Zhang et al, 2022.pdf](https://github.com/user-attachments/files/15829245/Zhang.et.al.2022.pdf)

**Comment by Snuffleupagus** [comment]
> The most effective way to implement this would probably be to see if the clickable link has the word "Fig" in it and change the behaviour accordingly.

First of all, please note that this isn't technically feasible since destinations in PDF documents are not the same thing as e.g. links in HTML files. Note that a PDF destination only contains information about the one or more of the: page-number, coordinates, and zoom-level; but has no access to the text-content in neither the clickable area nor at the location of the destination.

Secondly, assuming it was even possible, trying to match against a few hard-coded strings would get it wrong in many more cases than it'd get it "right" and this would just become a never-ending source of bugs. Furthermore, since pages can be (almost) arbitrarily long there's many existing PDF documents were ignoring the destination position would be really annoying.

---

All-in-all, the only thing that we could *potentially* do is offer a preference to ignore the destination position unconditionally. However, as mentioned, it's simply impossible to implement this like you hoped (i.e. with some kind of automatic detection).

**Comment by timvandermeij** [comment]
I agree with WONTFIX based on all information above. Ignoring the destination position unconditionally is in my opinion also bound to confuse users, even if it's a preference, because users might forget that they toggled it and file bug reports (and users might not know how to set such a preference anyway). Overall I also don't see how we can implement this in a good user-friendly way without unwanted side-effects unfortunately, so unless a really good idea is provided I don't think we can do this.

**Comment by anewuser** [comment]
Thank you for discussing and considering it.
